clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/HelixGroupRoutingStrategy.java

```diff
@@ -0,0 +1,109 @@
+  protected final InstanceHealthMonitor instanceHealthMonitor;
+  protected final HelixGroupStats helixGroupStats;
+
+  public HelixGroupRoutingStrategy(
+      InstanceHealthMonitor instanceHealthMonitor,
+      MetricsRepository metricsRepository,
+      String storeName) {
+    this(instanceHealthMonitor, new HelixGroupStats(metricsRepository, storeName));
+  }
+
+  HelixGroupRoutingStrategy(InstanceHealthMonitor instanceHealthMonitor, HelixGroupStats helixGroupStats) {
+    this.instanceHealthMonitor = instanceHealthMonitor;
+    this.helixGroupInfoAtomicReference.set(new HelixGroupInfo(Collections.emptyMap()));
+    this.helixGroupStats = helixGroupStats;
+  }
+
+  @Override
+  public String getReplicas(long requestId, int groupId, List<String> replicas) {
+    HelixGroupInfo helixGroupInfo = helixGroupInfoAtomicReference.get();
+    if (replicas.isEmpty() || helixGroupInfo.getHelixGroupInfoMap().isEmpty()) {
+      return null;
+    }
+    // select replicas from the selected group, going down the groups if more replicas are needed
+    int groupCnt = helixGroupInfo.getGroupIds().size();
+    List<Integer> groupIds = helixGroupInfo.getGroupIds();
+    Map<String, Integer> instanceToGroupIdMapping = helixGroupInfo.getHelixGroupInfoMap();
+    for (int i = 0; i < groupCnt; i++) {
+      int tmpGroupId = groupIds.get((i + groupId) % groupCnt);
+      for (String replica: replicas) {
+        if (instanceToGroupIdMapping.get(replica) == tmpGroupId && !instanceHealthMonitor.isInstanceBlocked(replica)
+            && instanceHealthMonitor.isInstanceHealthy(replica)) {
+          return replica;
+        }
+      }
+    }
+
+    return null;
+  }
+
+  public int getGroupCount() {
+    HelixGroupInfo helixGroupInfo = helixGroupInfoAtomicReference.get();
+    return helixGroupInfo.getGroupIds().size();
+  }
+
+  @Override
+  public void updateHelixGroupInfo(Map<String, Integer> instanceToHelixGroupIdMap) {
+    helixGroupInfoAtomicReference.set(new HelixGroupInfo(Collections.unmodifiableMap(instanceToHelixGroupIdMap)));
+  }
+
+  @Override
+  public int getHelixGroupId(long requestId, int groupIdForOriginalRequest) {
+    HelixGroupInfo helixGroupInfo = helixGroupInfoAtomicReference.get();
+    if (helixGroupInfo.getHelixGroupInfoMap().isEmpty()) {
+      throw new VeniceClientException("HelixGroupInfo is empty");
+    }
+    int newGroupId = (int) (requestId % helixGroupInfo.getGroupIds().size());
+    if (newGroupId == groupIdForOriginalRequest) {
+      newGroupId = (groupIdForOriginalRequest + 1) % helixGroupInfo.getGroupIds().size();
+    }
+    return newGroupId;
+  }
+
+  @Override
+  public boolean trackRequest(RequestContext requestContext) {
+    RequestType requestType = requestContext.getRequestType();
+    if (requestType.equals(RequestType.SINGLE_GET)) {
+      // So far, we mainly track multi-key requests here to monitor the group load.
```

Comment:
can you explain why are we just using the multi-key requests. Single get also could be keeping the instances busy right?

Reply:
Helix group based routing is mainly to scale large batch-get request and for single-get request, a normal round-robin should be good enough.
So far, we only track avg response time for all requests, so single-get request is normally very fast, so it can muddy the water.
But I do realize a problem if the customer only sends single-get request, no batch-get request, so I am going to add some randomness when picking up helix group id.